File: app/src/main/java/com/vincent/githubusersearch/network/DataCallbacks.java
package com.vincent.githubusersearch.network;

import android.util.Log;

import com.vincent.githubusersearch.AppController;
import com.vincent.githubusersearch.R;
import com.vincent.githubusersearch.callbacks.OnDataGetCallback;
import com.vincent.githubusersearch.model.ApiUrl;
import com.vincent.githubusersearch.model.ItemSearchResult;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.internal.EverythingIsNonNull;

public class DataCallbacks implements ApiUrl {

    private static final String TAG = "DataCallbacks";

    private static ApiInterface getApiInterface(String baseUrl) {
        return WebAgent.getRetrofit(baseUrl).create(ApiInterface.class);
    }

    private static<Item> void enqueue(Call<Item> call, final OnDataGetCallback<Item> callback) {
        Log.i(TAG, "Call URL: " + call.request().url().toString());

        call.enqueue(new Callback<Item>() {
            @EverythingIsNonNull
            @Override
            public void onResponse(Call<Item> call, Response<Item> response) {
                Log.d(TAG, "Call onResponse!!! "
                        + "\nMessage: " + response.message()
                        + " IsSuccessful: " + response.isSuccessful());

                if (response.isSuccessful()) {
                    callback.onDataGet(response.body());
                }
                else {
                    callback.onDataGetFailed(response.message());
                }
            }

            @EverythingIsNonNull
            @Override
            public void onFailure(Call<Item> call, Throwable t) {
                Log.e(TAG, "Call onFailure!!!\n" + t.getMessage());
                callback.onDataGetFailed(t.getMessage());
            }
        });
    }

    public static void getGitHubUsers(String keyword, int perPage, int page, OnDataGetCallback<ItemSearchResult> callback) {
        String clientId = AppController.Companion.getInstance().getString(R.string.github_client_id);
        String clientSecret = AppController.Companion.getInstance().getString(R.string.github_client_secret);

        Call<ItemSearchResult> call = getApiInterface(API_BASE_GITHUB_SEARCH).getGitHubUsers(keyword, perPage, page, clientId, clientSecret);

        enqueue(call, callback);
    }
}

File: app/src/main/java/com/vincent/githubusersearch/network/DataCallbacks.kt
package com.vincent.githubusersearch.network

import android.util.Log
import com.vincent.githubusersearch.AppController
import com.vincent.githubusersearch.R
import com.vincent.githubusersearch.callbacks.OnDataGetCallback
import com.vincent.githubusersearch.model.ApiUrl
import com.vincent.githubusersearch.model.ItemSearchResult
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.internal.EverythingIsNonNull

@Suppress("SameParameterValue")
object DataCallbacks {

    private const val TAG = "DataCallbacks"

    private fun getApiInterface(): ApiInterface {
        return getApiInterface(ApiUrl.API_BASE_GITHUB_SEARCH)
    }

    private fun getApiInterface(baseUrl: String): ApiInterface {
        return RetrofitAgent.getRetrofit(baseUrl).create(ApiInterface::class.java)
    }

    private fun <Item> enqueue(call: Call<Item>, callback: OnDataGetCallback<Item>) {
        Log.i(TAG, "Call URL: " + call.request().url().toString())

        call.enqueue(object : Callback<Item> {
            @EverythingIsNonNull
            override fun onResponse(call: Call<Item>, response: Response<Item>) {
                Log.d(TAG, "Call onResponse!!!\nMessage: ${response.message()} IsSuccessful: ${response.isSuccessful}")

                if (response.isSuccessful) {
                    callback.onDataGet(response.body())
                }
                else {
                    callback.onDataGetFailed(response.message())
                }
            }

            @EverythingIsNonNull
            override fun onFailure(call: Call<Item>, t: Throwable) {
                Log.e(TAG, "Call onFailure!!!\n${t.message}")
                callback.onDataGetFailed(t.message)
            }
        })
    }

    fun getGitHubUsers(keyword: String?, perPage: Int, page: Int, callback: OnDataGetCallback<ItemSearchResult>) {
        val clientId = AppController.instance.getString(R.string.github_client_id)
        val clientSecret = AppController.instance.getString(R.string.github_client_secret)

        val call = getApiInterface().getGitHubUsers(keyword, perPage, page, clientId, clientSecret)

        enqueue(call, callback)
    }
}